### PR TITLE
Add Median of Medians fallback to introselect

### LINKED ISSUE #102451: select_nth_unstable has quadratic worst-case time complexity; docs claim it should be linear
[The docs](https://doc.rust-lang.org/beta/std/primitive.slice.html#method.select_nth_unstable) claim that select_nth_unstable is "O(n) worst-case", but the implementation exhibits O(n<sup>2</sup>) time complexity for some inputs. I'm not sure if this is a documentation error or an implementation error, but it's definitely one or the other.

The current implementation is a fairly normal [Quickselect](https://en.wikipedia.org/wiki/Quickselect). Notably it only looks at a constant number of elements to determine a pivot, which _cannot_ guarantee linear time complexity for quickselect. [Median of medians](https://en.wikipedia.org/wiki/Median_of_medians) must be used instead, which is different from the "median of medians" referenced in the code.

For actual proof that pathological cases exist, here's a [playground link](https://play.rust-lang.org/?version=nightly&mode=release&edition=2021&gist=93aab67512b1eb4295cb2f886541424a). It generates close to the worst possible input for a fixed length, and compares the time for that against the time for a random input. For a length of ~128k (the highest I can get the playground to go before it times out), the pathological input is around 1000 times slower than the random input.

<details>
<summary>
Here's some graphs of runtime vs input length:
</summary>

time vs input length

time vs input length, log-log

Both show runtime in seconds vs slice length, with blue being random inputs and red being pathological inputs. The second is a log-log graph with some lines fit to the data. The slopes on the log-log graph are roughly 1.3 and 2.3 for the blue and red lines respectively.
</details>

To uphold the linear worst-case guarantee, select_nth_unstable would need to use [Median of medians](https://en.wikipedia.org/wiki/Median_of_medians). [Introselect](https://en.wikipedia.org/wiki/Introselect) is the hybrid version, and [Fast Deterministic Selection](https://arxiv.org/abs/1606.00484) is the most recent work I'm aware of that looks at optimizing median of medians.

If the worst-case guarantee is relaxed to an average-case guarantee (that's all that C++ gives for [nth_element](https://en.cppreference.com/w/cpp/algorithm/nth_element)), I think the docs should make it clear that the worst case is quadratic.

### PR DESCRIPTION
**Pull request by Sp00ph**
Fixes #102451.

This PR is a follow up to #106997. It adds a Fast Deterministic Selection implementation as a fallback to the introselect algorithm used by `select_nth_unstable`. This allows it to guarantee O(n) worst case running time, while maintaining good performance in all cases. 

This would fix #102451, which was opened because the `select_nth_unstable` docs falsely claimed that it had O(n) worst case performance, even though it was actually quadratic in the worst case. #106997 improved the worst case complexity to O(n log n) by using heapsort as a fallback, and this PR further improves it to O(n) (this would also make #106933 unnecessary). 
It also improves the actual runtime if the fallback gets called: Using a pathological input of size `1 << 19` (see the playground link in #102451), calculating the median is roughly 3x faster using fast deterministic selection as a fallback than it is using heapsort. 

The downside to this is less code reuse between the sorting and selection algorithms, but I don't think it's that bad. The additional algorithms are ~250 LOC with no `unsafe` blocks (I tried using unsafe to avoid bounds checks but it didn't noticeably improve the performance). 
I also let it fuzz for a while against the current `select_nth_unstable` implementation to ensure correctness, and it seems to still fulfill all the necessary postconditions.

cc @scottmcm who reviewed #106997

### CODE REVIEW

**Comment by Sp00ph** [comment]
Forgot to mention, on random input the fast deterministic select is a bit slower than the pdq based quickselect, so I didn't make `select_nth_unstable` always call the new algorithm and instead just used it it the introselect fallback.

**Comment by scottmcm** [comment]
Sorry, I'm going to flip this over to a real libs member since it's a bigger change than just a tweak.

r? rust-lang/libs

**Comment by scottmcm** [comment]
Ahh, my apologies, I read `cc` as `r?` 🤦

Putting it back,
r? @cuviper

**Comment by Sp00ph** [comment]
I tried also implementing the `ExpandPartition` algorithm from the paper that takes into account which region of the slice we already know to be partitioned, but it doesn't really lead to noticeable speedups. I think it's better to just use the already existing `partition` (like I did in this PR), as that already seems pretty optimized.

**Comment by Sp00ph** [comment]
Hmm, now that #106933 was merged, the API guarantees have been loosened to O(n log n) worst case runtime. I'm guessing we'd need an ACP to get it back down to O(n) worst case (and it's unclear to me if we'd even want to guarantee O(n)). Should I add note in the `/// ## Current implementation` section that states that the current algorithm (from this PR) always runs in linear time?

**Comment by orlp** [comment]
So first and foremost, the PR in its current form does not make the behavior linear, because the `limit` is set to `log(n)`, so in the worst case we're still `O(n log n)`.

Secondly, I think having the full Fast Deterministic Selection algorithm as a *fallback* algorithm does not make much sense. Either it should be run from start to finish, or we should reduce it to just the `median_of_ninthers` approach for its linearity. The fallback algorithm only really gets triggered by an incredibly unlucky input pattern or malicious input, so it does not make sense to have a lot of code baked into the executable to run our fallback approach at maximum speed.

**Comment by Sp00ph** [comment]
> So first and foremost, the PR in its current form does not make the behavior linear, because the `limit` is set to `log(n)`, so in the worst case we're still `O(n log n)`.

I had naively thought that the same limiting strategy used for introsort would also work for introselect, but yeah makes sense that it doesn't. Does this mean that to guarantee linear runtime we need a constant limit on the iterations?

> Secondly, I think having the full Fast Deterministic Selection algorithm as a *fallback* algorithm does not make much sense. Either it should be run from start to finish, or we should reduce it to just the `median_of_ninthers` approach for its linearity. The fallback algorithm only really gets triggered by an incredibly unlucky input pattern or malicious input, so it does not make sense to have a lot of code baked into the executable to run our fallback approach at maximum speed.

In my tests, just using fast deterministic selection unconditionally was a bit slower than the current median-of-3 quickselect on random inputs, so I didn't want to do that. I also tried to just use median of medians first, but that was slower than using heapsort as a fallback. I haven't tried just using median of ninthers yet, I'll have to see if that still gets similar speedups to the full fast deterministic select.

**Comment by Sp00ph** [comment]
I suppose it could do something like `let limit = cmp::min(ilog2(len), SOME_UPPER_BOUND)` which would still mean O(1) iterations. Would probably have to figure out a good value for that cutoff point though.

**Comment by orlp** [comment]
> I suppose it could do something like `let limit = cmp::min(ilog2(len), SOME_UPPER_BOUND)` which would still mean O(1) iterations. Would probably have to figure out a good value for that cutoff point though.

I think you should do what I did in pdqsort originally: track whether the partition was 'good' or 'bad' by some threshold (I used n / 8 = 12.5% quantile for the pivot position as the cutoff), and then and only then increment the counter. This way you can put the counter relatively low without affecting performance for normal code while quickly switching to the fallback for malicious/unlucky inputs.

**Comment by workingjubilee** [comment]
> In my tests, just using fast deterministic selection unconditionally was a bit slower than the current median-of-3 quickselect on random inputs, so I didn't want to do that. I also tried to just use median of medians first, but that was slower than using heapsort as a fallback. I haven't tried just using median of ninthers yet, I'll have to see if that still gets similar speedups to the full fast deterministic select.

Performance on perfectly random inputs is slightly overrated. What about nonrandom inputs, like tokenized data from Wikipedia or such?

**Comment by Sp00ph** [comment]
I currently don't have access to a powerful machine, but I can try to set up a somewhat comprehensive benchmark suite once I do to compare heapsort fallback vs median-of-ninthers fallback vs just fast deterministic select.

**Comment by cuviper** [comment]
Sorry, I haven't had time to dig into this - let me roll a new reviewer...

r? libs

**Comment by Sp00ph** [comment]
I still need to change some stuff about the PR so

@rustbot author

**Comment by Sp00ph** [comment]
Not yet, I'll see if I can get around to it some time soon.

**Comment by Sp00ph** [comment]
I wrote a little benchmark suite testing the current selection implementation, quickselect with median of medians as fallback and pure fast deterministic select on various slice lengths and input shapes and quickselect with median of medians fallback was consistently the fastest, so I changed it to that now. If anyone is interested I can upload the benchmark itself too. The algorithm should now also actually be O(n) worst case since I made the quickselect iteration limit constant now.

File: library/core/src/slice/select.rs [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+
+    let (left, right) = v.split_at_mut(index);
+    let (pivot, right) = right.split_at_mut(1);
+    let pivot = &mut pivot[0];
+    (left, pivot, right)
+}
+
+/// helper function used to find the index of the min/max element
+/// using e.g. `slice.iter().enumerate().min_by(from_is_less(&mut is_less)).unwrap()`
+fn from_is_less<T>(
```
[the remark this answers is not in the document]
Comment (Amanieu) [reply library/core/src/slice/select.rs:142]: > It seems like the code could be made more readable by just having `min_index` and `max_index` helper functions that return an `Option<usize>`.

I don't think this has been addressed yet.

[the remark this answers is not in the document]
Comment (Sp00ph) [reply library/core/src/slice/select.rs]: Yeah I kinda forgot about it, should be fixed now though.

**Comment by Amanieu** [comment]
This can be submitted as a separate PR which we will run an FCP on, since this is a new stable guarantee. No ACP is required.

**Pull request merged**